Add blog post about I18n (#248)
infinum/eightshift-docs: website/blog/2024-02-01-making-your-project-multilingual.md

## The I18n class
The easiest way to add I18n support to your project is by using our WP CLI boilerplate command:
```bash
wp boilerplate create i18n
```

This command generated a new class inside the `src/I18n` folder. This class instructs WordPress to look for translations in `src/I18n/languages` with the textdomain defined as your project name. The next step is generating .po and .mo files that are used for translation.

## Generating .pot file
You can create a .pot file by using WP CLI. Run the following command in your project root:
```bash
wp i18n make-pot
```

Alternatively, you can use [Poedit](https://poedit.net/) to generate the .pot file and generate translations from it later.

## Translating with Poedit
Once you have the .pot file, you can use Poedit to generate .po and .mo files that are used for translating hardcoded strings in your project. When generating the .po and .mo files, you can choose for which locale you're creating the translation. For example, if you are creating a translation for the German language, your files will be named de_DE.po and de_DE.mo.

> After generating the files, you should go to **Translation -> Properties** and go to **Sources Paths** tab. For the Base path, you should set your theme folder. In Excluded paths, you can add folders like `node_modules`, `vendor`, and `public`.

Reviewer: ```suggestion
After generating the files, go to **Translation -> Properties** and navigate to the **Sources Paths** tab. Set the _Base path_ to the theme folder path. In _Excluded paths_ you can add folders like `node_modules`, `vendor`, and `public` to exclude external packages.
```

> In the **Sources keywords** tab you can set additional functions which you use in your project for translations. Some of the most commonly used include:

Reviewer: ```suggestion
In the _Sources keywords_ tab you can set additional functions for use in your project for translations. Commonly used functions are:
```

> - _e
> - __
> - esc_html__
> - esc_html_e
> - esc_attr__
> - esc_html_x
> - _n

Reviewer: ```suggestion
- `_e` for translating a string and echoing it
- `__` for returning a translated string
- `esc_html__` for returning a translated string which is escaped in a way it's safe to use within HTML
- `esc_html_e` for echoing a translated string which is escaped in a way it's safe to use within HTML
- `esc_attr__` for returning a translated string which is escaped in a way it's safe to use within an attribute
- `esc_html_x` for returning a translated string which is escaped in a way it's safe to use within HTML, with a _gettext_ context
- _n
```

> > If you're missing a string in your .po file, be sure to check which function is used for translation for that string and if that function is added to Sources keywords.

Reviewer: ```suggestion
> If you're missing a string in your `.po` file be sure to check which function is used for translation for that string, and that the function is added to _Sources keywords_.
```

> After updating the settings, click on the Update from source code option to get the updated list of strings to translate.

Reviewer: ```suggestion
After updating the settings, click on _Update from source code_ option to get the updated list of strings to translate.
```

> The translation process is simple. The left column represents the source text, and the right column the translation. When you have finished translating the strings, copy the .po and .mo files in the `src/I18n/languages` folder.

Reviewer: ```suggestion
The translation process is simple. The left column represents the source text, and the right column the translation. When you have finished translating the strings, copy the `.po` and `.mo` files to the `src/I18n/languages` folder.
```

## JS translations

> The process of translating strings in JS has some extra steps. In order to translate the strings in JS (e.g. block editor), you will have to generate a .json file with translations. To do this, navigate to your `src/I18n/languages` folder and use the following WP-CLI command:

Reviewer: ```suggestion
The process of translating strings in JS has a couple of extra steps.

In order to translate strings in JS (e.g. Block editor strings), you will have to generate translation file. To do this, navigate to your `src/I18n/languages` folder and use the following WP-CLI command:
```

```bash
wp i18n make-json <po-file> --no-purge
```

> This will generate a .json file for each JS file you have. This command extracts the strings from your .po files, so you'll already have the translations added. The `--no-purge` flag is used to keep the existing translations in your .po file.

Reviewer: ```suggestion
This will generate a `.json` file for each JS file present. The strings are extracted from `.po` files, so you'll already have the translations added. The `--no-purge` flag is used to keep the existing translations in the `.po` file.
```

> The method used for setting the script translations is `setScriptTranslations()` in the `I18n` class.

Reviewer: ```suggestion
The method used for setting the script translations is `setScriptTranslations()` from the `I18n` class.
```

> The default way this works in our setup is that you need to have a single .json file for all your JS translations. You can either modify this method to read from all JSON files, or merge all the JSON files in a single one.

Reviewer: ```suggestion
The default way this works in Eightshift DevKit is that you need to have a single `.json` file with all the JS translations. If needed, you can either modify this method to read from multiple files, or just merge all the `.json` files into one.
```

> If you're going with the default method, your file should follow this naming structure:

Reviewer: ```suggestion
If using the default setup (everything in one file) follow this naming structure: `{textdomain}-{locale}-{handle}.json`.
```

> `{textdomain}-{locale}-{handle}.json`. For example, if your textdomain is `project-name` and your locale is `de_DE`, your file should be named `project-name-de_DE-project-name-block-editor-scripts.json`.

Reviewer: ```suggestion

For example, if your _textdomain_ is `project-name` and your locale is `de_DE`, your file should be named `project-name-de_DE-project-name-block-editor-scripts.json`.
```

> > The block-related translations depend on the language the user has set in WP Admin.

Reviewer: ```suggestion
> The block-related translations depend on the language the user has set in WP admin.
```

## Enabling languages and content translation

> To have the option of switching between languages and translate the content you've written in WordPress, your best bet is to use a plugin. Here are a few plugins that we actively use or recommend:
> - **WPML** - one of the most popular plugins on the market. It is a paid plugin but offers a lot of advanced options
> - **Polylang** - free plugin, but also has a paid Pro version

Reviewer: ```suggestion
If the website itself needs to support content in multiple languages, a plugin is a good option.

 The most common multi-language plugins are:
- **WPML** - one of the most popular plugins on the market. It is a paid plugin, but offers a lot of advanced options.
- **Polylang** - a free plugin (also has a paid _Pro_ version).
```

> Please keep in mind that this is just a summary of plugins we recommend and that you should thoroughly research various multilingual plugins and decide which would be the best for your project.

Reviewer: ```suggestion
> Explore other options as well, you might find a plugin that is a better fit for your project than WPML or Polylang.
```

> Most of the translation work will be through the editor since you'll need to translate the content on posts and pages. The best approach for this is to use one of the multilingual plugins like WPML.

Reviewer: ```suggestion
Most of the translation work will be done through the editor, since you'll need to translate the content on posts and pages.
```

## Additional resources

> There are a lot of details regarding Internationalization (I18n) and Localization (L10n), so it's impossible to cover everything in a single blog post. If you wish to know about the core I18n functionality or a bit more about how it is being used in the Eightshift Development kit, here are a few resources which you may find interesting:

Reviewer: ```suggestion
Internationalization (_I18n_) and Localization (_L10n_) are very broad topics, so it's impossible to cover everything in a single blog post.

If you wish to know about the core I18n functionalities, or a bit more about how it is used in the Eightshift DevKit, here are a few resources which you may find interesting:
```

After generating the files, go to **Translation -> Properties** and navigate to the **Sources Paths** tab. Set the _Base path_ to the theme folder path. In _Excluded paths_ you can add folders like `node_modules`, `vendor`, and `public` to exclude external packages.

In the _Sources keywords_ tab you can set additional functions for use in your project for translations. Commonly used functions are:
- `_e` for translating a string and echoing it
- `__` for returning a translated string
- `esc_html__` for returning a translated string which is escaped in a way it's safe to use within HTML
- `esc_html_e` for echoing a translated string which is escaped in a way it's safe to use within HTML
- `esc_attr__` for returning a translated string which is escaped in a way it's safe to use within an attribute
- `esc_html_x` for returning a translated string which is escaped in a way it's safe to use within HTML, with a _gettext_ context

> - _n

Reviewer 2: What about this line?

> - _n

Author: Soon(TM) 😁 I still haven't re-requested review, I just committed your suggestions

> - _n

Reviewer 2: Excuse my premature review

> The block-related translations depend on the language the user has set in WP admin.

## Enabling languages and content translation
If the website itself needs to support content in multiple languages, a plugin is a good option.

 The most common multi-language plugins are:
- **WPML** - one of the most popular plugins on the market. It is a paid plugin, but offers a lot of advanced options.
- **Polylang** - a free plugin (also has a paid _Pro_ version).

> Explore other options as well, you might find a plugin that is a better fit for your project than WPML or Polylang.

Most of the translation work will be done through the editor, since you'll need to translate the content on posts and pages.

## Additional resources
Internationalization (_I18n_) and Localization (_L10n_) are very broad topics, so it's impossible to cover everything in a single blog post.

If you wish to know about the core I18n functionalities, or a bit more about how it is used in the Eightshift DevKit, here are a few resources which you may find interesting:
- [WordPress Codex - I18n for WordPress Developers](https://codex.wordpress.org/I18n_for_WordPress_Developers)

> - [Eightshift Development kit documentation - Tips & Tricks](https://eightshift.com/docs/basics/tips-tricks/#internationalization-i18n-and-localization-l10n)

Reviewer 2: ```suggestion
- [Eightshift Development kit documentation - Tips & Tricks](https://eightshift.com/docs/basics/tips-tricks/#internationalization-i18n-and-localization-l10n)
- [Infinum WordPress Handbook - Localization](https://infinum.com/handbook/wordpress/translations/localization)
```
